### PR TITLE
Fix server sundry

### PR DESCRIPTION
**Pull request by tacaswell**

### CODE REVIEW

File: caproto/_data.py [hunk 1]
```diff
@@ -249,6 +271,7 @@ def get_dbr_data(self, type_):
     async def set_dbr_data(self, data, data_type, metadata):
         '''Set data from DBR metadata/values'''
         self.value = self.fromtype(values=data, data_type=data_type)
+        self.timestamp = time.time()
```
Comment (danielballan) [inline caproto/_data.py]: Any thoughts on how/whether we should ensure that this timestamp is consistent with the timestamp optionally passed in through `metadata`?

File: caproto/_data.py [hunk 2]
```diff
@@ -38,7 +41,26 @@ def _convert_char_values(values, to_dtype, string_encoding, enum_strings):
 
 
 def _convert_string_values(values, to_dtype, string_encoding, enum_strings):
-    if to_dtype in native_int_types or to_dtype in native_float_types:
+    if to_dtype == ChType.ENUM:
```
Comment (danielballan) [inline caproto/_data.py]: Grosssssssss. :-D

We'll come back for this.

File: caproto/_data.py (hunk 1 above)
Comment (klauer) [inline caproto/_data.py]: I think we probably should just check if it's present in `metadata`, otherwise use `time.time()`?

Comment (tacaswell) [inline caproto/_data.py]: In the cases I cared about `metadata` was always `None`....

**Comment by tacaswell** [comment]
rebased to drop the cherry-picked commits.

File: caproto/_data.py (hunk 1 above)
Comment (danielballan) [reply caproto/_data.py]: (I'm addressing this in #90.)

**Comment by danielballan** [comment]
Test failures reproducible locally:

```
tests/test_examples.py::test_curio_server_with_caget[char-ChannelType.CHAR] FAILED
tests/test_examples.py::test_curio_server_with_caget[char-ChannelType.STS_CHAR] FAILED
tests/test_examples.py::test_curio_server_with_caget[char-ChannelType.TIME_CHAR] FAILED
tests/test_examples.py::test_curio_server_with_caget[char-ChannelType.GR_CHAR] FAILED
tests/test_examples.py::test_curio_server_with_caget[char-ChannelType.CTRL_CHAR] FAILED
```

(There are sometimes other random failures that seem to be a separate issue.)

This PR fixes a bunch of critical bugs. Unclear whether it also introduces new ones or merely exposes existing bugs that weren't getting properly tested before.

**Comment by danielballan** [comment]
Tests pass. I'll merge this for the critical fixes (e.g. missing `await`s), open an issue to revisit CHAR length, and carry on with _data.py improvements in #90 as remarked above, back in June.

**Comment by klauer** [comment]
👏 🎆

**Pull request merged**